### DISCUSSION TITLE
Override 422 UnprocessableEntity response

### DISCUSSION THREAD

**Discussion by atistler** [luolingchun/flask-openapi3#88] [General]
Currently pydantic validation errors are rendered something like this:

**422 UnprocessableEntity**
```json
[
	{
		"loc": [
			"job_type"
		],
		"msg": "field required",
		"type": "value_error.missing"
	},
	{
		"loc": [
			"details",
			"action"
		],
		"msg": "field required",
		"type": "value_error.missing"
	},
	{
		"loc": [
			"details",
			"options"
		],
		"msg": "field required",
		"type": "value_error.missing"
	}
]
```

Is there anyway to override this behavior and make the response body customizable specifically for 422's when the body is invalid.

**Comment by atistler** [comment]
I tried to specify a "422" in the response but that does not seem to make a difference:

```
        @self.app.post(
            "/v2/jobs",
            summary="Create job",
            operation_id="createJob",
            description="Creates a job",
            responses={"201": response.Job, "403": response.UnauthorizedError, "422": response.UnprocessableEntityError, "409": response.ConflictError},
            tags=[jobs_tag],
        )
        def create_job(body: CreateJobBody):
```

**Comment by luolingchun** [comment]
@atistler Why you need override response 422？Can you write a simple demo.

Pydantic validation errors will be raised when validate pydantic model, refer to https://github.com/luolingchun/flask-openapi3/blob/master/flask_openapi3/request.py#L134-L134

**Comment by atistler** [comment]
Because all other http errors in my application are structured in the following way:
```
{
   code: str,
   message: str
}
```

I would like to maintain that consistency.  I guess i could write a decorator wrapping the app.post() that catches the pydantic validation error.  I would seem that flask-openapi3 shouldn't make assumptions about how the 422 should be rendered as an http response though.

**Comment by luolingchun** [comment]
We have drafted a PR ([Be able to change 422 validation errors to other http response status code](https://github.com/luolingchun/flask-openapi3/pull/83)) to discuss the 422 UnprocessableEntity response, So we can merge and solve this problem.

My idea is to add `validation_error_status`, `validation_error_model`, `validation_error_callback` to `OpenAPI`:

- `validation_error_status`: the status code of the response
- `validation_error_model`: response model for OpenAPI Specification
- `validation_error_callback`: response callback, the return format corresponds to the `validation_error_model`

```python
from pydantic import ValidationError, BaseModel


class ValidationErrorModel(BaseModel):
     code: str,
     message: str

def validation_error_callback(e: ValidationError):
    validation_error_object = ValidationErrorModel(code="400", mmessage=e.json())
    response = make_response(validation_error_object.json())
    response.headers["Content-Type"] = "application/json"
    response.status_code = getattr(current_app, "validation_error_status", 422)
    return response


app = OpenAPI(__name__, validation_error_status=422, validation_error_model=ValidationErrorModel, validation_error_callback=validation_error_callback)
```

And then change the logic of [request](https://github.com/luolingchun/flask-openapi3/blob/master/flask_openapi3/request.py#L134-L134).


```python
try:
    # Validate header, cookie, path, and query parameters
    if header:
        _do_header(header, func_kwargs)
    if cookie:
        _do_cookie(cookie, func_kwargs)
    if path:
        _do_path(path, path_kwargs, func_kwargs)
    if query:
        _do_query(query, func_kwargs)
    if form:
        _do_form(form, func_kwargs)
    if body:
        _do_body(body, func_kwargs)
except ValidationError as e:
    # Create a JSON response with validation error details
    validation_error_callback = getattr(current_app, "validation_error_callback", None)
    if validation_error_callback:
        response = validation_error_callback(e)
    else:
        response = make_response(e.json())
        response.headers["Content-Type"] = "application/json"
        response.status_code = getattr(current_app, "validation_error_status", 422)
    return response
```
